packages/runtime/runtime-utils/src/generated/serializer.ejs

```diff
@@ -0,0 +1,238 @@
+                    ? ${varName}.url
+                    : generateHandleContextPath(${varName}.url, this.context);
+
+                return new RemoteFluidObjectHandle(absolutePath, this.root);
+            }
+        `);
+    }
+
+    /**
+     * Emits 'recursivelyEncode()' and 'recursivelyDecode()'.
+     */
+    function recursively(name, fn, args) {
+        const callArgs = ["input: any", ...args].join(", ");
+        const recurArgs = ["value", ...(args.map(arg => arg.split(":")[0]))].join(", ");
+        const fnName = name.split("recursively")[1].toLowerCase();
+
+        return trim(`
+            // Invoked for non-null objects to recursively ${fnName} references to IFluidHandles.
+            // Clones as-needed to avoid mutating the 'input' object.  If no IFluidHandes are present,
+            // returns the original 'input'.
+            private ${name}(${callArgs}) {
+                // Note: Caller is responsible for ensuring that 'input' is a non-null object.
+
+${indent(fn, 4)}
+
+                // Otherwise descend into the object graph looking for IFluidHandle instances.
+                // eslint-disable-next-line @typescript-eslint/ban-types
+                let clone: object | undefined;
+                for (const key of Object.keys(input)) {
+                    const value = input[key];
+                    // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
+                    if (!!value && typeof value === "object") {
+                        // Note: Except for IFluidHandle, 'input' must not contain circular references (as object must
+                        //       be JSON serializable.)  Therefore, guarding against infinite recursion here would only
+                        //       lead to a later error when attempting to stringify().
+                        const replaced = this.${name}(${recurArgs});
+
+                        // If the 'replaced' object is different than the original 'value' then the subgraph contained one
+                        // or more handles.  If this happens, we need to return a clone of the 'input' object where the
+                        // current property is replaced by the 'replaced' value.
+                        if (replaced !== value) {
+                            // Lazily create a shallow clone of the 'input' object if we haven't done so already.
+                            clone = clone ?? (Array.isArray(input)
+                                ? [...input]
+                                : { ...input });
+
+                            // Overwrite the current property 'key' in the clone with the 'replaced' value.
+                            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+                            clone![key] = replaced;
+                        }
+                    }
+                }
+                return clone ?? input;
+            }
+        `);
+    }
+_%>
+
```

Comment:
The remainder is emitted as-is, except for the <%- -%> blocks, which call the above functions to inline code templates.